server/proxy/proxy.go

```diff
@@ -178,3 +187,67 @@ func (p *Proxy) upstreamForApp(app *apps.App) (upstream.Upstream, error) {
 		return nil, utils.NewInvalidError("not a valid app type: %s", app.AppType)
 	}
 }
+
+func (p *Proxy) CleanUserCallContext(userID string, cc *apps.Context) (*apps.Context, error) {
+	cc = &apps.Context{
+		ContextFromUserAgent: cc.ContextFromUserAgent,
+	}
+
+	var postID, channelID, teamID string
+
+	switch {
+	case cc.PostID != "":
+		postID = cc.PostID
+
+		post, err := p.mm.Post.GetPost(postID)
+		if err != nil {
+			return nil, err
+		}
+
+		channelID = post.ChannelId
+
+		_, err = p.mm.Channel.GetMember(channelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, channelID)
+		}
+
+		c, err := p.mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
+
+	case cc.ChannelID != "":
+		channelID = cc.ChannelID
+
+		_, err := p.mm.Channel.GetMember(cc.ChannelID, userID)
```

Comment:
I both a `ChannelID` and a `PostID` are provided `GetMember` is done twice. Can we drop one?

Reply:
@hanzei That's not true, because this case statement won't run if the post id was provided